### PR TITLE
Continuous palette

### LINKED ISSUE #1695: contour colormap
the curves have new colors (#1436) but not the contours

see openturns/lib/src/Base/Func/EvaluationImplementation.cxx:
 current.setColor(Contour::ConvertFromHSV((360.0 * i / levels.getDimension()), 1.0, 1.0)); 
same code is used in Distribution, Field, CovModel, BoxCoxFactory

I suggest to take viridis only (matplotlib default linear colormap) from:
https://github.com/yuki-koyama/tinycolormap/blob/master/include/tinycolormap.hpp

### PR DESCRIPTION
**Pull request by mbaudin47**
This is a follow up to #2475.

This PR creates two static methods:
- `ot.Drawable.BuildGrayScalePalette` : a grayscale palette for black & white prints,
- `ot.Drawable.BuildViridisPalette` : a viridis palette for plots requiring a continuous set of colors.

Indeed, the Viridis colormap is usable for color blind people, which is not the case for some other colormaps (see [^nunez2018]).

This PR also fixes the use of a rainbow palette which is used in some `draw()` methods:
- `EvaluationImplementation.draw()`
- `CovarianceModelImplementation.draw()`
- `FieldImplementation.draw()`
- `BoxCoxFactory::buildWithGraph()`

The change involves replacing the call to the `Curve::ConvertFromHSV` method to the viridis palette of colors.

For example, the following code (from [plot_estimate_non_parametric_copula.html](http://openturns.github.io/openturns/latest/auto_data_analysis/estimate_dependency_and_copulas/plot_estimate_non_parametric_copula.html)):
```python
distribution = ot.BernsteinCopulaFactory().build(sample)
graph = distribution.drawPDF()
```
changes the graph:

image

Moreover, it is now easy to set the color scale depending on a specific parameter. I customized the [plot_viscous_fall_field_function_connection.html](http://openturns.github.io/openturns/latest/auto_functional_modeling/field_functions/plot_viscous_fall_field_function_connection.html) and selected a color based on the maximum height achieved by each trajectory.
The next cell computes the maximum height of each trajectory. Then the `argsort` method computes the indices of the cells which sort the sample.
```python
maximumHeight = ot.Sample.BuildFromPoint([outputSample[i].getMax()[0] for i in range(size)])
index = maximumHeight.argsort()
palette = ot.Drawable.BuildGrayScalePalette(size)  # or palette = ot.Drawable.BuildViridisPalette(size)
graph.setColors(palette[index])
```

image

This PR will close #1695.

[^nunez2018]: Nuñez, J. R., Anderton, C. R., & Renslow, R. S. (2018). Optimizing colormaps with consideration for color vision deficiency to enable accurate interpretation of scientific data. PloS one, 13(7). https://arxiv.org/pdf/1712.01662.pdf

### CODE REVIEW

File: python/src/DrawableImplementation_doc.i.in [hunk 1]
```diff
@@ -204,14 +204,73 @@ linearily depending on the 'Drawable-DefaultPalettePhase' key of the
...
+
+Parameters
+----------
+n : int :math:`n > 0` and :math:`n < 10`
+    Number of colors needed.
+
+Returns
+-------
+listColors : :class:`~openturns.Description`
+    List of n color codes defined according to the viridis palette.
```
Comment (mbaudin47) [inline python/src/DrawableImplementation_doc.i.in:226]:
```suggestion
    List of *n* color codes defined according to the viridis palette.
```

File: python/src/DrawableImplementation_doc.i.in [hunk 2]
```diff
@@ -204,14 +204,73 @@ linearily depending on the 'Drawable-DefaultPalettePhase' key of the
...
+
+Parameters
+----------
+n : int :math:`n > 0` and :math:`n < 10`
+    Number of colors needed.
+
+Returns
+-------
+listColors : :class:`~openturns.Description`
+    List of n color codes defined according to the gray scale palette.
```
Comment (mbaudin47) [inline python/src/DrawableImplementation_doc.i.in:255]:
```suggestion
    List of *n* color codes defined according to the gray scale palette.
```

**Comment by mbaudin47** [comment]
@jschueller , @josephmure : I think that this is ready for a review. Can you please tell me if there is anything to improve?

File: lib/src/Base/Func/EvaluationImplementation.cxx [hunk 3]
```diff
@@ -432,14 +432,15 @@ Graph EvaluationImplementation::draw(const UnsignedInteger firstInputMarginal,
   isoValues.buildDefaultLabels();
   const Point levels(isoValues.getLevels());
   const Description labels(isoValues.getLabels());
+  const Description palette = DrawableImplementation::BuildViridisPalette(levels.getDimension());
```
Comment (jschueller) [inline lib/src/Base/Func/EvaluationImplementation.cxx:435]: why hardcode viridis now that you're adding everything needed to switch colormaps ?

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 4]
```diff
@@ -1611,6 +1620,49 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+  if (!(size > 0)) throw InvalidArgumentException(HERE) << "Error: the size must be > 0, but is " << size;
+  Description palette(size);
+  UnsignedInteger firstValue = 255 - std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-WhiteMargin")),
+    lastValue = std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-BlackMargin"));
+  if (size > 1)
+  {
+    Scalar step = (lastValue - (Scalar)firstValue) / (size - 1);
+    for (UnsignedInteger i = 0; i < size; i++)
+    {
+      UnsignedInteger val = std::min(255UL, static_cast<UnsignedInteger>(round(firstValue + i * step)));
```
Comment (jschueller) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1657]: const

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 5]
```diff
@@ -1611,6 +1620,49 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+  return palette;
+}
+
+/* Build GrayScale palette
+   Distributes colors evenly ensuring that the first and last colors match the white and black margins */
+Description DrawableImplementation::BuildGrayScalePalette(const UnsignedInteger size)
+{
+  if (!(size > 0)) throw InvalidArgumentException(HERE) << "Error: the size must be > 0, but is " << size;
+  Description palette(size);
+  UnsignedInteger firstValue = 255 - std::min(255UL, ResourceMap::GetAsUnsignedInteger("Drawable-WhiteMargin")),
```
Comment (jschueller) [inline lib/src/Base/Graph/DrawableImplementation.cxx:1650]: const Scalar ?

File: lib/src/Base/Stat/FieldImplementation.cxx [hunk 6]
```diff
@@ -594,8 +591,7 @@ Graph FieldImplementation::drawMarginal(const UnsignedInteger index,
     else
     {
       const UnsignedInteger size = marginalValues.getSize();
-      Description palette(size);
-      for (UnsignedInteger i = 0; i < size; ++i) palette[i] = Curve::ConvertFromHSV((270.0 * (size - i - 1)) / size, 1.0, 1.0);
+      const Description palette = DrawableImplementation::BuildViridisPalette(size);
```
Comment (jschueller) [inline lib/src/Base/Stat/FieldImplementation.cxx:594]: same

File: lib/src/Base/Func/EvaluationImplementation.cxx (hunk 3 above)
Comment (mbaudin47) [reply lib/src/Base/Func/EvaluationImplementation.cxx:435]: The idea was to replace a code using a rainbow colormap to something more appropriate. I admit that replacing a hard-coded palette with another hard-coded palette was not the best option. I see at least to methods to implement your idea:

1) We might use the existing `BuildDefaultPalette` method
```cpp
const Description palette = DrawableImplementation::BuildDefaultPalette(levels.getDimension(), );
```
In this case, the default palette will be Tableau. The user can customize this parameter by setting the 'Drawable-DefaultPaletteName' ResourceMap key. 

2) Assuming that we create a new `BuildPalette(size, paletteName)` where `paletteName` is a `string`, then we could use a default `ResourceMap` key:
```cpp
const Description paletteName = ResourceMap::GetAsString("Evaluation-Palette");
const Description palette = DrawableImplementation::BuildPalette(levels.getDimension(), paletteName);
```
The default value of this key might be Viridis. The user can customize this by setting the "Evaluation-Palette" ResourceMap key.

@jschueller : Which option would fit best in your opinion?

File: lib/src/Uncertainty/Algorithm/Transformation/ProcessTransformation/BoxCoxFactory.cxx [hunk 7]
```diff
@@ -432,6 +432,7 @@ BoxCoxTransform BoxCoxFactory::buildWithGraph(const Sample & sample,
   for (UnsignedInteger i = 0; i < npts; ++i) lambdaValues(i, 0) = xMin + i * (xMax - xMin) / (npts - 1.0);
   Point lambdaMarginal(1, 0.0);
   Sample logLikelihoodValues(npts, 1);
+  const Description palette = DrawableImplementation::BuildViridisPalette(dimension);
```
[the remark this answers is not in the document]
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/Transformation/ProcessTransformation/BoxCoxFactory.cxx:435]: Fixed at 4af90e7ca876f7c97ecb67acb3f69ff4958b780f

File: lib/src/Base/Graph/DrawableImplementation.cxx [hunk 8]
```diff
@@ -1611,6 +1620,49 @@ Description DrawableImplementation::BuildTableauPalette(const UnsignedInteger si
...
+  {
+    InitializeValidParameterList();
+    IsFirstInitialization = false;
+  }
+  if (!(size > 0)) throw InvalidArgumentException(HERE) << "Error: the size must be > 0, but is " << size;
+  Description palette(size);
+  palette[size - 1] = ViridisColors[ViridisColors.getSize() - 1];
+  if (size > 1)
+  {
+    Scalar step = (ViridisColors.getSize() - 1.) / (size - 1);
```
[the remark this answers is not in the document]
Comment (mbaudin47) [reply lib/src/Base/Graph/DrawableImplementation.cxx]: Fixed at 4af90e7ca876f7c97ecb67acb3f69ff4958b780f

File: lib/src/Base/Graph/DrawableImplementation.cxx (hunk 5 above)
Comment (mbaudin47) [reply lib/src/Base/Graph/DrawableImplementation.cxx]: Fixed at 4af90e7ca876f7c97ecb67acb3f69ff4958b780f

File: lib/src/Base/Graph/DrawableImplementation.cxx (hunk 4 above)
Comment (mbaudin47) [reply lib/src/Base/Graph/DrawableImplementation.cxx]: Fixed at 4af90e7ca876f7c97ecb67acb3f69ff4958b780f

File: python/doc/examples/graphs/plot_graphs_loglikelihood_contour.py [hunk 9]
```diff
@@ -161,31 +162,26 @@ def logLikelihood(X):
...
+ot.ResourceMap.SetAsUnsignedInteger("Drawable-BlackMargin", 20)
+ot.ResourceMap.SetAsUnsignedInteger("Drawable-WhiteMargin", 50)
+palette = ot.Drawable.BuildGrayScalePalette(len(levels))
 # Create the drawables list, appending each contour with its own color
-drawables = list()
-for i in range(len(levels)):
-    contour.setLevels([levels[i]])
+drawables = []
+for level in levels:
+    contour.setLevels([level])
```
Comment (josephmure) [inline python/doc/examples/graphs/plot_graphs_loglikelihood_contour.py:184]: We will be able to do all this more easily with #2587

**Review by josephmure — COMMENTED** [review]
I am still unconvinced that the manual introduction of the viridis and grayscale palettes is necessary.

Some of the proposed improvements are already tackled by #2587, or else could be more easily tackled if based on #2587.

Regarding the example you wrote in the original post of this PR, the matplotlib lines required to manually do similar things are not very numerous, and could be freely used in the examples.

**Comment by josephmure** [comment]
Here is an example which chooses colors of curves based on their maximum height (it is based on the example from the original post of this PR) using matplotlib only.

## Prepare the levels the colors will be based on

```python
import matplotlib
maximumHeight = [outputSample[i].getMax()[0] for i in range(size)]
margin = (max(maximumHeight) - min(maximumHeight)) / size # white margin necessary with grayscale
norm = matplotlib.colors.Normalize(vmax = max(maximumHeight) + margin)
color_levels = norm(maximumHeight)
```

## Create a viridis palette based on these levels

```python
viridis = matplotlib.cm.get_cmap('viridis')
viridis_palette = [matplotlib.colors.to_hex(c) for c in viridis(color_levels)]
```

## Alternatively, create a grayscale palette
```python
grayscale = matplotlib.cm.get_cmap('gray')
grayscale_palette = [matplotlib.colors.to_hex(c) for c in grayscale(color_levels)]
```

## Set the colors (viridis or grayscale)
```python
graph.setColors(viridis_palette) # graph.setColors(grayscale_palette)
```

## Result

image
image

**Comment by josephmure** [comment]
Note that we can also easily add a colorbar with matplotlib:

```python
view = viewer.View(graph)
fig = view.getFigure()
axes = view.getAxes()
fig.colorbar(matplotlib.cm.ScalarMappable(norm=norm, cmap=viridis), ax=axes[0])
```

Result:
image

File: lib/src/Uncertainty/Algorithm/Transformation/ProcessTransformation/BoxCoxFactory.cxx [hunk 10]
```diff
@@ -441,7 +442,7 @@ BoxCoxTransform BoxCoxFactory::buildWithGraph(const Sample & sample,
       logLikelihoodValues(i, 0) = boxCoxOptimization(lambdaMarginal)[0];
     }
     Curve curve(lambdaValues, logLikelihoodValues);
-    curve.setColor(Curve::ConvertFromHSV((360.0 * d) / dimension, 1.0, 1.0));
+    curve.setColor(palette[d]);
```
[the remark this answers is not in the document]
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/Transformation/ProcessTransformation/BoxCoxFactory.cxx:445]: I cannot see any example in the doc to see what the result looks like.

**Comment by mbaudin47** [comment]
In terms of customization, you are right: Matplotlib does the job. Let's push it further: what would we need _any_ color map after all? One of the reasons is to get usable plots "plug and play" that is that we can generally use without any customization. The "problem" is that the library is in C++. Hence, we cannot generate the colormap right in the source code. This means that we must implement some colormaps that generally fit the purpose. It does not imply that we should implement all of them: the user can customize it afterwards using Matplotlib. But must implement _some of them_.

In order to see the effects of the changes in the source code you pointed, I considered them one after the other.




| Before                                   | After                                    |
| ---------------------------------------- | ---------------------------------------- |
|  Pasted image 20240405225635 | Pasted image 20240405225625  |
| Pasted image 20240405225645 | Pasted image 20240405225706 |
| Pasted image 20240405225655 | Pasted image 20240405225711 |
**Table 1.** The result of `Field.draw()`from [plot_field_manipulation.html](http://openturns.github.io/openturns/latest/auto_probabilistic_modeling/stochastic_processes/plot_field_manipulation.html).


| Before                                   | After                                    |
| ---------------------------------------- | ---------------------------------------- |
|  |  |
|  |  |
**Table 2.** The result of `BoxCoxFactory().buildWithGraph` from [BoxCoxFactory.html](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.BoxCoxFactory.html)

| Before                                   | After                                    |
| ---------------------------------------- | ---------------------------------------- |
|  |  |
**Table 3.** The example from [plot_graphs_loglikelihood_contour.html](http://openturns.github.io/openturns/latest/auto_graphs/plot_graphs_loglikelihood_contour.html)

**Comment by josephmure** [comment]
This is a really detailed compilation! I will try to comment each of your Before/After examples:

## The arrow representation of a field on 2-d mesh and 2-d values

OpenTURNS creates the arrows as Curve objects and assigns a color to each "curve" (ie arrow) based on the arrow length. This handcrafted representation of a field exists because OpenTURNS has no internal C++ representation of the Matplotlib [Quiver](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.quiver.html) object, which is dedicated to field representation.
I think that the correct way to draw such graphs would be to implement a C++ container for the quiver arguments. One of them is the colormap (just like for contour plots). Notice however that by default quiver uses only one color for all arrows, and I think that is for the best: there is no guarantee the color should be chosen based on the vector norm. Indeed, in the picture you show, this choice seems rather arbitrary.
Therefore, my short term proposition would be to change the "draw" method of Fields in order to use only one color for all arrows. If the user wants to color them using a color palette, it is always possible to do it afterwards like in my script in the posts above.

## Marginal field (no interpolation)

Again, I think the problem comes from the fact we have no C++ representation of the appropriate Matplotlib object, which is a [Triangulation](https://matplotlib.org/stable/api/tri_api.html#matplotlib.tri.Triangulation) to be used within the [tripcolor method](https://matplotlib.org/stable/gallery/images_contours_and_fields/tripcolor_demo.html). Unfortunately I cannot think of any short-term solution in this case.

## Marginal field (with interpolation)

In this case, I think the method should be rewritten to use the Contour class. Then it will be able to benefit from #2587.

## Box Cox log-likelihood example

I agree with you: we need to use the Tableau palette.

## Contour example plotting a log-likelihood function

This example is [already modified](https://output.circle-artifacts.com/output/job/b45c8df7-b837-453d-9d99-7ae519772ca5/artifacts/0/html/auto_graphs/plot_graphs_loglikelihood_contour.html?highlight=loglikelihood) by #2587 (which also adds a colorbar).

image

**Comment by josephmure** [comment]
> My conclusion is as follows.
>
> * If we do not implement the Viridis colormap, then we cannot produce plots that use it right out of the library. Hence the user _will_ have to customize it when the data represents a value. This is because none of the palette currently available in the library has a scale which fits the purpose.

This is not completely true IMO: contour plots produced by draw methods will use the viridis colormap out of the box after #2587 is merged. Other plots could benefit as well if we create appropriate classes to contain the arguments (e.g. Triangulation / Quiver).

> * When [Filled contour/colobar #2587](https://github.com/openturns/openturns/pull/2587) will be available, then we will be able to use the Viridis colormap _name_ from the C++. But we still won't be able to compute the list of associated colors, so that we can set them in a set of curves. Can you see a way to create a set of Viridis colored set of curves after [Filled contour/colobar #2587](https://github.com/openturns/openturns/pull/2587) is merged? I cannot see any (please provide your analysis of this topic if you have an idea in mind).

I have a feeling the usecases are likely to actually be cases where what is missing is an appropriate Drawable which could be mapped to a Matplotlib object that uses a colormap. In the case of curves maybe we need a CurveArray class similar to PolygonArray? Then the viewer could automatically do the Python lines of code I showed.

> * The PR [Filled contour/colobar #2587](https://github.com/openturns/openturns/pull/2587) will simplify the process to build a _filled_ contour plot. But it will not change the process to create an unfilled one.

Actually, it will by default produce unfilled contours with viridis (see the picture above).

> * The use of the colormap for the BoxCox factory is not appropriate in my opinion: the tableau palette will fit better.

Agreed.

**Comment by mbaudin47** [comment]
@josephmure : I agree with most of your comments. So we can put this PR on break, and see how this goes after #2587 .

**Comment by jschueller** [comment]
joseph: feature redundant with  color maps

**Comment by mbaudin47** [comment]
After a discussion during the cotech, we decided to close this PR.

**Pull request closed without merging**